platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/MapRenderer.java

```diff
@@ -121,12 +122,38 @@ private native void nativeInitialize(MapRenderer self,
   private void updateFps() {
     frames++;
     long currentTime = System.nanoTime();
-    double fps = 0;
-    if (currentTime - timeElapsed >= 1) {
-      fps = frames / ((currentTime - timeElapsed) / 1E9);
+    long timeCost = currentTime - timeElapsed;
+
+    if (timeCost >= 1) {
+      double fps = frames / ((timeCost) / 1E9);
+      if (fps > preferredFramesPerSecond) {
+        try {
+          long millis = (long) (timeCost / 1E6 * (fps - preferredFramesPerSecond) / fps);
+          Thread.sleep(millis);
```

Comment:
on which thread this would be executed? what's the perf impact of this?

Reply:
It's on GLThread, by doing this, the fps will be lower.